{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO\n",
    "from Bio.AlignIO import MafIO\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we will parse the maf alignment into AlignIO. This will return a generator object for all the alignments in the maf file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse the maf alignment\n",
    "# THIS FILE CANNOT BE COMMITTED DUE TO BEING LARGE\n",
    "# alignment = AlignIO.parse('../data/small_subset/chr21.maf', 'maf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SingleLetterAlphabet() alignment with 4 rows and 4490 columns\n",
      "gagtcatcagaacgaaacagaacagaacagggagtttcgcaatg...T-G Gorilla_gorilla_gorilla.chr21\n",
      "tagtcatcagaacgaaacagaacagaacagggagtttcgcaatg...TGG Homo_sapiens.chr21\n",
      "tagtcatcagaatgaaacagaacagaacagggagtttcgcaatg...TGG Pan_paniscus.CM003404.1\n",
      "tagtcatcagaacgaaacagaacagaacagggagtttcgcaatg...TGG Pongo_abelii.CM009283.2\n"
     ]
    }
   ],
   "source": [
    "alignment = AlignIO.parse('../results/small_subset/chr21_subset.maf', 'maf')\n",
    "print(next(alignment))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we want to know the start and end coordinates of the maf file we can use the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def start_end(alignment, target_seqname, window_size):\n",
    "    \"\"\"\n",
    "    This function returns a list containing the start\n",
    "    and end coordinates of a sliced multiple sequence \n",
    "    alignment from a generator object given a target\n",
    "    species and chromosome in the form of species.chromosome.\n",
    "    The slicing is done in base of the user-specified window \n",
    "    size.\n",
    "    \"\"\"\n",
    "    # Create empty list of start coordinates\n",
    "    start = []\n",
    "    # Create empty list of sizes\n",
    "    size = []\n",
    "    # For each alignment in the file\n",
    "    for ali in alignment:\n",
    "        # For each record in the alignment\n",
    "        for record in ali:\n",
    "            # Find the reference sequence\n",
    "            if record.name == target_seqname:\n",
    "                # Record the start position\n",
    "                start.append(record.annotations['start'])\n",
    "                # Record the size\n",
    "                size.append(record.annotations['size'])\n",
    "    # Create empty list of coordinates\n",
    "    coord_lst = []\n",
    "    # Save first start coordinate\n",
    "    st = start[0]\n",
    "    # Save first size\n",
    "    sz = size[0]\n",
    "    # For each index\n",
    "    for i in range(1, len(start)):\n",
    "        # If the index is not the last one\n",
    "        if i < (len(start)-1):\n",
    "            # If the size is smaller than the window size\n",
    "            if sz+size[i] < window_size:\n",
    "                # Update the accumulated size\n",
    "                sz += size[i]\n",
    "            # If the size is larger than the window size\n",
    "            else:\n",
    "                # Append start and end coordinates\n",
    "                coord_lst.append((st, start[i]+size[i]-1))\n",
    "                # Update the accumulated size and start position\n",
    "                st = start[i+1]\n",
    "                sz = size[i]\n",
    "        else:\n",
    "            coord_lst.append((st, start[i]+size[i]-1))\n",
    "    return(coord_lst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(19999946, 20596357), (20596359, 21171293), (21171294, 21762193), (21762194, 22407773), (22407774, 23024473), (23024487, 23722282), (23723096, 24456841), (24460262, 25004156)]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignment = AlignIO.parse('../results/small_subset/chr21_subset.maf', 'maf')\n",
    "slicing_lst = start_end(alignment, 'Homo_sapiens.chr21', 500000)\n",
    "print(slicing_lst)\n",
    "len(slicing_lst)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now index the maf file using the following function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "idx = MafIO.MafIndex(\"../results/chr21_subset.mafindex\", '../results/small_subset/chr21_subset.maf', \"Homo_sapiens.chr21\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we want to retrieve a 50 kb window, we can do so with the search method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = idx.search([20000000], [20050000])\n",
    "AlignIO.write(results, '../results/small_subset/chr21_subset.fa', \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also save each of the alignments separately. For each of the coalHMM runs we will have a set of individual fasta files, whose name will be encoded as `chr21_subset_{number of run}_{fasta identifier}.fa`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "for num, sli in enumerate(slicing_lst):\n",
    "    results = idx.search([sli[0]], [sli[1]])\n",
    "    for i, j in enumerate(results):\n",
    "        AlignIO.write(j, '../results/small_subset/individual_alignments/chr21_subset_{}_{}.fa'.format(num, i), \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And we can save the coordinates easily:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_coord_df(align):\n",
    "    \"\"\"\n",
    "    Function for getting a data frame with the species, the\n",
    "    start coordinates and a binary vector where 1 means a nt\n",
    "    and 0 means a gap.\n",
    "    \"\"\"\n",
    "    # Create an empty dataframe\n",
    "    df = pd.DataFrame(columns = ['file', 'species', 'chr', 'start', 'gaps'])\n",
    "    # For each of the alignments\n",
    "    for i, align in enumerate(results):\n",
    "        # Create empty dictionary\n",
    "        dct = {'species':[], 'chr':[], 'start':[],'gaps':[]}\n",
    "        # Write individual FASTA file\n",
    "        AlignIO.write(align, '../results/small_subset/individual_alignments/chr21_subset{}.fa'.format(i), \"fasta\")\n",
    "        # For each of the records\n",
    "        for record in align:\n",
    "            # Retrieve species\n",
    "            dct['species'].append(record.name.split('.')[0])\n",
    "            # Retrieve chromosome/contig\n",
    "            dct['chr'].append('.'.join(record.name.split('.')[1:]))\n",
    "            # Retrieve start coordinate\n",
    "            dct['start'].append(record.annotations['start'])\n",
    "            # Retrieve gaps encoded in a binary format\n",
    "            dct['gaps'].append(''.join([str(0) if n=='-' else str(1) for n in record.seq]))\n",
    "        # Convert dictionary to data frame\n",
    "        file_df = pd.DataFrame.from_dict(dct)\n",
    "        # Insert column mapping to the file\n",
    "        file_df.insert(0, 'file', i, True)\n",
    "        # Append rows to overall data frame\n",
    "        df = df.append(file_df)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>species</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>gaps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6586419</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>19999946</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20071554</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7175441</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6590887</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20004414</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20076022</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7179789</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6595123</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20008672</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20080282</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7184043</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6597056</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20010607</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20081983</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7185990</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6599046</td>\n",
       "      <td>1100000000000000000000111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20012587</td>\n",
       "      <td>1100000000000000000000111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20083941</td>\n",
       "      <td>0111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7187978</td>\n",
       "      <td>1100000000000000000000111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6603148</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20016692</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20088064</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7192081</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6605912</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20019467</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20090829</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7194868</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6607854</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20021414</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20092772</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7196803</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6611843</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20025428</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20096787</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7200814</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6614699</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20028237</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20099527</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7203288</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6616765</td>\n",
       "      <td>1111111111111111111111111111111111111111001111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20030302</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20101590</td>\n",
       "      <td>1111111111111000111111111111111111111111001111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7205350</td>\n",
       "      <td>1111111111111111111111111111111111111111001111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6620251</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20033796</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20105073</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7208823</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6621750</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20035307</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20106582</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7210278</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6624950</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20038516</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20110104</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7213448</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14</td>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6629964</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>chr21</td>\n",
       "      <td>20043536</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>CM003404.1</td>\n",
       "      <td>20115119</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>Pongo_abelii</td>\n",
       "      <td>CM009283.2</td>\n",
       "      <td>7218442</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  file                  species         chr     start  \\\n",
       "0    0  Gorilla_gorilla_gorilla       chr21   6586419   \n",
       "1    0             Homo_sapiens       chr21  19999946   \n",
       "2    0             Pan_paniscus  CM003404.1  20071554   \n",
       "3    0             Pongo_abelii  CM009283.2   7175441   \n",
       "0    1  Gorilla_gorilla_gorilla       chr21   6590887   \n",
       "1    1             Homo_sapiens       chr21  20004414   \n",
       "2    1             Pan_paniscus  CM003404.1  20076022   \n",
       "3    1             Pongo_abelii  CM009283.2   7179789   \n",
       "0    2  Gorilla_gorilla_gorilla       chr21   6595123   \n",
       "1    2             Homo_sapiens       chr21  20008672   \n",
       "2    2             Pan_paniscus  CM003404.1  20080282   \n",
       "3    2             Pongo_abelii  CM009283.2   7184043   \n",
       "0    3  Gorilla_gorilla_gorilla       chr21   6597056   \n",
       "1    3             Homo_sapiens       chr21  20010607   \n",
       "2    3             Pan_paniscus  CM003404.1  20081983   \n",
       "3    3             Pongo_abelii  CM009283.2   7185990   \n",
       "0    4  Gorilla_gorilla_gorilla       chr21   6599046   \n",
       "1    4             Homo_sapiens       chr21  20012587   \n",
       "2    4             Pan_paniscus  CM003404.1  20083941   \n",
       "3    4             Pongo_abelii  CM009283.2   7187978   \n",
       "0    5  Gorilla_gorilla_gorilla       chr21   6603148   \n",
       "1    5             Homo_sapiens       chr21  20016692   \n",
       "2    5             Pan_paniscus  CM003404.1  20088064   \n",
       "3    5             Pongo_abelii  CM009283.2   7192081   \n",
       "0    6  Gorilla_gorilla_gorilla       chr21   6605912   \n",
       "1    6             Homo_sapiens       chr21  20019467   \n",
       "2    6             Pan_paniscus  CM003404.1  20090829   \n",
       "3    6             Pongo_abelii  CM009283.2   7194868   \n",
       "0    7  Gorilla_gorilla_gorilla       chr21   6607854   \n",
       "1    7             Homo_sapiens       chr21  20021414   \n",
       "2    7             Pan_paniscus  CM003404.1  20092772   \n",
       "3    7             Pongo_abelii  CM009283.2   7196803   \n",
       "0    8  Gorilla_gorilla_gorilla       chr21   6611843   \n",
       "1    8             Homo_sapiens       chr21  20025428   \n",
       "2    8             Pan_paniscus  CM003404.1  20096787   \n",
       "3    8             Pongo_abelii  CM009283.2   7200814   \n",
       "0    9  Gorilla_gorilla_gorilla       chr21   6614699   \n",
       "1    9             Homo_sapiens       chr21  20028237   \n",
       "2    9             Pan_paniscus  CM003404.1  20099527   \n",
       "3    9             Pongo_abelii  CM009283.2   7203288   \n",
       "0   10  Gorilla_gorilla_gorilla       chr21   6616765   \n",
       "1   10             Homo_sapiens       chr21  20030302   \n",
       "2   10             Pan_paniscus  CM003404.1  20101590   \n",
       "3   10             Pongo_abelii  CM009283.2   7205350   \n",
       "0   11  Gorilla_gorilla_gorilla       chr21   6620251   \n",
       "1   11             Homo_sapiens       chr21  20033796   \n",
       "2   11             Pan_paniscus  CM003404.1  20105073   \n",
       "3   11             Pongo_abelii  CM009283.2   7208823   \n",
       "0   12  Gorilla_gorilla_gorilla       chr21   6621750   \n",
       "1   12             Homo_sapiens       chr21  20035307   \n",
       "2   12             Pan_paniscus  CM003404.1  20106582   \n",
       "3   12             Pongo_abelii  CM009283.2   7210278   \n",
       "0   13  Gorilla_gorilla_gorilla       chr21   6624950   \n",
       "1   13             Homo_sapiens       chr21  20038516   \n",
       "2   13             Pan_paniscus  CM003404.1  20110104   \n",
       "3   13             Pongo_abelii  CM009283.2   7213448   \n",
       "0   14  Gorilla_gorilla_gorilla       chr21   6629964   \n",
       "1   14             Homo_sapiens       chr21  20043536   \n",
       "2   14             Pan_paniscus  CM003404.1  20115119   \n",
       "3   14             Pongo_abelii  CM009283.2   7218442   \n",
       "\n",
       "                                                gaps  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  \n",
       "0  1100000000000000000000111111111111111111111111...  \n",
       "1  1100000000000000000000111111111111111111111111...  \n",
       "2  0111111111111111111111111111111111111111111111...  \n",
       "3  1100000000000000000000111111111111111111111111...  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  \n",
       "0  1111111111111111111111111111111111111111001111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111000111111111111111111111111001111...  \n",
       "3  1111111111111111111111111111111111111111001111...  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  \n",
       "0  1111111111111111111111111111111111111111111111...  \n",
       "1  1111111111111111111111111111111111111111111111...  \n",
       "2  1111111111111111111111111111111111111111111111...  \n",
       "3  1111111111111111111111111111111111111111111111...  "
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = idx.search([20000000], [20050000])\n",
    "get_coord_df(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And as a sanity check we can also save the end coordinates from the alignment and the end coordinates calculated by summing the binary vector of gaps to the start coordinates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "dct2 = {}\n",
    "results = idx.search([20000000], [20050000])\n",
    "\n",
    "for i, align in enumerate(results):\n",
    "    dct2[i] = {'end':[], 'end2':[]}\n",
    "    AlignIO.write(align, '../results/small_subset/individual_alignments/chr21_subset{}.fa'.format(i), \"fasta\")\n",
    "    for record in align:  \n",
    "        dct2[i]['end'].append(record.annotations['start']+record.annotations['size'])\n",
    "        dct2[i]['end2'].append(record.annotations['start']+sum([0 if n=='-' else 1 for n in record.seq]))\n",
    "\n",
    "for i in dct2:\n",
    "    print(dct2[i]['end'] == dct2[i]['end2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
